### PR TITLE
Feature/137 housing wanted create endpoint

### PR DESCRIPTION
**Pull request by AlainNshimirimana**

### CODE REVIEW

File: lib/housing-wanted/housing-wanted.models.js [hunk 1]
```diff
@@ -29,5 +29,18 @@ module.exports = {
...
     return HousingWanted;
   },
-
-};
+  id: Joi.object().keys({
+    id: Joi.string().guid()
+  }),
+  api: Joi.object().keys({
+    title: Joi.string().required(),
+    bedrooms: Joi.number().integer().min(0).max(9),
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.models.js]: max should be 10 as per property.models

File: lib/housing-wanted/housing-wanted.models.js [hunk 2]
```diff
@@ -29,5 +29,18 @@ module.exports = {
...
+    bathrooms: Joi.number().precision(2).min(1).max(9),
+    contact: Joi.string(),
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.models.js]: I think contact should be required as well so users can be contacted if someone is interested in their post.

File: lib/housing-wanted-location/housing-wanted-location.models.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+  db: (sequelize, Sequelize) => {
+    const HousingWantedLocation = sequelize.define('HousingWantedLocation', {
+      
+    });
+
+    return HousingWantedLocation;
+  },
+  
+};
```
Comment (zaclittleberry) [inline lib/housing-wanted-location/housing-wanted-location.models.js]: hrrmmm. still no newline at the end of files. I thought we fixed this in your editor config. I wonder what's happening.

File: index.js [hunk 4]
```diff
@@ -83,6 +83,7 @@ module.exports = (async() => {
       modelsFile = require(`./lib/${mod}/${mod}.models.js`);
       if (modelsFile.db) {
         let model = modelsFile.db(sequelize, Sequelize);
+        //console.log(model.name);
```
Comment (zaclittleberry) [inline index.js]: ideally this would be removed/file would be completely reverted back to its previous state so no unnecessary changes were made.

File: lib/housing-wanted/housing-wanted.controllers.js [hunk 5]
```diff
@@ -1,7 +1,118 @@
...
+          }
+        }
+
+        // bathrooms validation
+        let hasBathrooms = false;
+        let bathrooms = Number(request.payload.bathrooms);
+        if (bathrooms) {
+          hasBathrooms = true;
+        }
+
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.controllers.js]: is superseded by lines 48-50. please remove

File: lib/housing-wanted/housing-wanted.controllers.js [hunk 6]
```diff
@@ -1,7 +1,118 @@
...
+        let housingWantedObject = {
+          AuthorId: account.id,
+          title: request.payload.title,
+          bedrooms: bedrooms,
+          bathrooms: bathrooms,
+          contact: request.payload.contact,
+          body: request.payload.body,
+          status: request.payload.status,
+          details: request.payload.details,
+        };
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.controllers.js]: Building the object this way is really only suitable when all the fields are required/guaranteed to be there, because it makes the assumption they are. For example, this assigns bedrooms and bathrooms, but they aren't required, which results in them being Number(undefined) which = NaN, which will fail on db insert. It also sets AuthorId to account.id regardless of whether AuthorId was submitted or not. 

The preferred way of doing this would be to create the housingWantedObject, and assign each property in a conditional that checks the appropriate thing. eg:
if (hasBathrooms) { housingWantedObject.bathrooms = bathrooms; }
if (request.AuthorId) { housingWantedObject.AuthorId = request.AuthorId; } else { housingWantedObject.AuthorId = account.id; }

File: lib/housing-wanted/housing-wanted.controllers.js [hunk 7]
```diff
@@ -1,7 +1,118 @@
...
+        for (let locationId of request.payload.LocationIds){
+          try {
+            await models.HousingWantedLocation.create({
+              HousingWantedId: housingWantedInstance.id,
+              LocationId: locationId
+            });
+          } catch (error) {
+            console.log(error);
+          }
+        }
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.controllers.js]: should be wrapped in a conditional checking if LocationIds were set on the payload, since they aren't guaranteed to be.

**Review by zaclittleberry — CHANGES_REQUESTED** [review]

File: lib/housing-wanted-location/housing-wanted-location.models.js (hunk 3 above)
Comment (AlainNshimirimana) [reply lib/housing-wanted-location/housing-wanted-location.models.js]: did you want me to add a new line at the end of the file?

File: lib/housing-wanted/housing-wanted.controllers.js [hunk 8]
```diff
@@ -63,27 +60,55 @@ module.exports = (models) => {
...
+        let housingWantedObject = {};
+        if (request.AuthorId){
+          housingWantedObject.AuthorId = request.AuthorId;
+        }
+        else{
+          housingWantedObject.AuthorId = account.id;
+        }
+
+        if (request.title){
+          housingWantedObject.title = request.title;
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.controllers.js]: missing .payload on request

File: lib/housing-wanted/housing-wanted.controllers.js [hunk 9]
```diff
@@ -63,27 +60,55 @@ module.exports = (models) => {
...
+        if (hasBedrooms){
+          housingWantedObject.bedrooms = bedrooms;
+        }
+
+        if (hasBathrooms){
+          housingWantedObject.bathrooms = bathrooms;
+        }
+
+        if (request.contact){
+          housingWantedObject.contact = request.contact;
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.controllers.js]: missing .payload on request

File: lib/housing-wanted/housing-wanted.controllers.js [hunk 10]
```diff
@@ -63,27 +60,55 @@ module.exports = (models) => {
...
+        }
+
+        if (request.body){
+          housingWantedObject.body = request.body;
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.controllers.js]: missing .payload on request

File: lib/housing-wanted/housing-wanted.controllers.js [hunk 11]
```diff
@@ -63,27 +60,55 @@ module.exports = (models) => {
...
+        }
+
+        if (request.payload.status){
+          housingWantedObject.status = request.status;
+        }
+
+        if (request.payload.details){
+          housingWantedObject.details = request.details;
```
Comment (zaclittleberry) [inline lib/housing-wanted/housing-wanted.controllers.js]: missing .payload on request

**Review by zaclittleberry — CHANGES_REQUESTED** [review]
in the change to incrementally building the data object, the .payload sub-object was dropped from the request object statememts.

File: lib/housing-wanted/housing-wanted.controllers.js [hunk 12]
```diff
@@ -1,7 +1,133 @@
...
+      postHousingWanted: async function(request, h){
+        /*
+          # Require Account for submitting fields other than address
+        */
+        let payload = request.payload;
+        let account;
+        try {
+         account = await accountLib.getAccount(request.auth.credentials);
+        } catch (e) {
+         throw Boom.badImplementation('Error during getAccount(request.auth.credentials). ' + e);
```
Comment (VitaRox) [inline lib/housing-wanted/housing-wanted.controllers.js]: this should be changed to `throw Boom.badImplementation('Error during getAccount(request.auth.credentials). ', e);`

Comment (zaclittleberry) [reply lib/housing-wanted/housing-wanted.controllers.js]: Yes! Thank you! Done, so I'm resolving this conversation.
